File: 04-structural-patterns/01-oo-decorator.py
from datetime import datetime

# Base Logger Interface
class Logger:
    def log(self, message: str):
        raise NotImplementedError()

# Concrete Component
class ConsoleLogger(Logger):
    def log(self, message: str):
        print(message)

# Base Decorator
class LoggerDecorator(Logger):
    def __init__(self, logger: Logger):
        self._logger = logger

    def log(self, message: str):
        self._logger.log(message)

# Concrete Decorators
class TimestampLogger(LoggerDecorator):
    def log(self, message: str):
        timestamped = f"{datetime.now()} - {message}"
        super().log(timestamped)

class LevelLogger(LoggerDecorator):
    def __init__(self, logger: Logger, level: str):
        super().__init__(logger)
        self._level = level

    def log(self, message: str):
        leveled = f"[{self._level.upper()}] {message}"
        super().log(leveled)

# Usage
logger = ConsoleLogger()
logger = TimestampLogger(logger)
logger = LevelLogger(logger, "debug")

logger.log("Something happened")

File: 04-structural-patterns/01-decorator/01-oo-decorator.go
package main

import (
	"fmt"
	"time"
)

// Component
type Logger interface {
	Log(message string)
}

// Concrete Component
type ConsoleLogger struct{}

func (ConsoleLogger) Log(message string) {
	fmt.Println(message)
}

// Base Decorator
type LoggerDecorator struct {
	logger Logger
}

func (d LoggerDecorator) Log(message string) {
	d.logger.Log(message)
}

// Timestamp Decorator
type TimestampLogger struct {
	LoggerDecorator
}

func (t TimestampLogger) Log(message string) {
	msg := fmt.Sprintf("%s - %s", time.Now().Format(time.RFC3339), message)
	t.logger.Log(msg)
}

// Level Decorator
type LevelLogger struct {
	LoggerDecorator
	level string
}

func (l LevelLogger) Log(message string) {
	msg := fmt.Sprintf("[%s] %s", l.level, message)
	l.logger.Log(msg)
}

func main() {
	var logger Logger = ConsoleLogger{}
	logger = TimestampLogger{LoggerDecorator{logger}}
	logger = LevelLogger{LoggerDecorator{logger}, "INFO"}

	logger.Log("App started")
}
